### PR TITLE
Remove zero_grad call in LightningOptimizer

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

In LightningOptimizer the `optimizer.zero_grad()`, shouldn't be there since it's already called within `training_loop.zero_grad_handler` after calling `on_before_zero_grad` hook. ATM it's called but with zero gradients of the parameters.

Also found 1 more bug. While doing `trainer.test`, optimizers are initialized, which they shouln't.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
I agree zero grad should not be inside our optimizer step. 
We should try to track zero grad and step calls with mocks in tests :)

**Comment by rohitgr7** [comment]
> We should try to track zero grad and step calls with mocks in tests :)

agreed +1

**Comment by rohitgr7** [comment]
@tchaton can you review the changes? will update other tests too after that if required.

**Comment by tchaton** [comment]
Hey @rohitgr7 @awaelchli @SeanNaren 

I disagree sightly with this design choice. 

Here are my reasons:

1 . Actually`zero_grad_handler` is used only for automatic optimization only and it is there only as a legacy support for model.optimizer_step
    
```
    def zero_grad_handler(self, batch_idx, optimizer, opt_idx):
        if self.automatic_optimization:
            # hook
            self.on_before_zero_grad(optimizer)
            optimizers = enumerate([optimizer])
        else:
            optimizers = []

        for idx, optimizer in optimizers:
            self.optimizer_zero_grad(batch_idx, optimizer, opt_idx)
```

2. Each optimizers should possibly have their own accumulated gradient logic.
 
Currently, in LightningOptimizer, the default is Trainer(accumulated_grad_batches=x).
It can be overridden by doing `optimizer.accumulated_grad_batches=y` or be used UDF logic such as
`optimizer.step(make_optimizer_step=self.trainer.batch_idx % 4 == 0)`.

We don't want our user to ever call `zero_grad`. The purpose of Lightning is take care of all those small details. 
If people use random logic which we can't predict, then `zero_grad` needs to called after `optimizer.step()`.

However, I agree we can make a small change as `zero_grad_handler` would be redundant if not.

```
        if make_optimizer_step:
            if trainer.on_tpu:
                with trainer.profiler.profile(profiler_name):
                    xm.optimizer_step(optimizer, optimizer_args={'closure': closure, **kwargs})

            elif trainer.amp_backend is not None:
                trainer.precision_connector.backend.optimizer_step(
                    trainer, optimizer, closure)

            else:
                with trainer.profiler.profile(profiler_name):
                    optimizer.step(closure=closure, *args, **kwargs)
        else:
            # make sure to call optimizer_closure when accumulating
            with trainer.profiler.profile("closure"):
                with trainer.train_loop.block_ddp_sync_behaviour():
                    closure()
         
         # zero_grad_handler would do this.
          if not automatic_optimization:
                 optimizer.zero_grad()

```

Finally, we currently have too many ways to perform `optimizer_step`, and LightningOptimizer should be the class to handle this logic.

I hope it makes sense. 

Best regards,
Thomas Chaton.

**Comment by tchaton** [comment]
Hey @SeanNaren @awaelchli @rohitgr7,

Here is a WIP for uniformizing the automatic and manual logic: https://github.com/PyTorchLightning/pytorch-lightning/pull/4984

Best regards,
Thomas Chaton.

**Comment by awaelchli** [comment]
> We don't want our user to ever call zero_grad

That's very unfortunate for so many reasons :(
